pyproject.toml

```diff
@@ -17,18 +17,33 @@
 [build-system]
 build-backend = 'mesonpy'
 requires = [
-    "meson-python>=0.15.0",
-    "Cython>=3.0.8",        # when updating version, also update check in meson.build
-    "pybind11>=2.13.2",     # when updating version, also update check in scipy/meson.build
-    "pythran>=0.14.0",
+    # The upper bound on meson-python is pre-emptive only (looser
+    # on purpose, since chance of breakage in 0.18/0.19 is low with
+    # 0.17.1 working at time of writing)
+    "meson-python>=0.15.0,<0.20.0",
+    # we need at least Cython 3.1.0a1 for free-threaded CPython
+    # 3.13 wheel builds
+    "Cython>=3.1.0a1,<3.2.0; python_version=='3.13t'",        # when updating version, also update check in meson.build
+    # for other CPython versions, the regular pre-emptive
+    # Cython version bounds policy applies
+    "Cython>=3.0.8,<3.1.0; python_version!='3.13t'",
+    # The upper bound on pybind11 is pre-emptive only
+    "pybind11>=2.13.2,<2.14.0",     # when updating version, also update check in scipy/meson.build
+    # The upper bound on pythran is pre-emptive only; 0.17.0
+    # is released/working at time of writing, and 0.18.x may
+    # have desirable free-threaded CPython 3.13 support
+    "pythran>=0.14.0,<0.19.0",
 
     # numpy requirement for wheel builds for distribution on PyPI - building
     # against 2.x yields wheels that are also compatible with numpy 1.x at
     # runtime.
     # Note that building against numpy 1.x works fine too - users and
     # redistributors can do this by installing the numpy version they like and
     # disabling build isolation.
-    "numpy>=2.0.0rc1",
+    "numpy>=2.0.0,<2.5; python_version!='3.13t'",
+    # free-threaded CPython 3.13 support requires more
+    # recent NumPy release at build time
+    "numpy>=2.1.3,<2.5; python_version=='3.13t'",
```

Comment:
Again, these won't work... probably just the usual one above and purge the Python version checks + disable build isolation for free-threaded..

Reply:
Same, changing to a code comment would be good.